File: ccc21s5.cpp
#include <bits/stdc++.h>
using namespace std;


// template <typename T>
// void pv(const vector<T>& v) {
// 	for (const T& x : v)
// 		cout << x << " ";
// 	cout << "\n";
// }

// gcds
namespace st1 {
	int n;
	vector<int> t;

	void init(int sz) {
		n = sz;
		t.resize(n*4+1, 1);
	}

	void update(int x, int y, int z, int l, int r, int i) {
		if (l >= x && r <= y) {
			t[i] = lcm(t[i], z);
		}
		else if (x <= r && y >= l) {
			int m = (l + r) / 2;
			update(x, y, z, l  , m, i*2  );
			update(x, y, z, m+1, r, i*2+1);
		}
	}
	void update(int x, int y, int z) {
		update(x-1, y-1, z, 0, n-1, 1);
	}
}

// seq
namespace st2 {
	int n;
	vector<int> t;
	vector<bool> s;

	void init(int sz) {
		n = sz;
		t.resize(n*4+1);
		s.resize(n*4+1);
	}

	void build(int l, int r, int i, int v = 1) {
		v = lcm(v, st1::t[i]);
		if (l == r) {
			t[i] = v;
			s[i] = true;
		}
		else {
			int m = (l + r) / 2;
			build(l  , m, i*2  , v);
			build(m+1, r, i*2+1, v);
			t[i] = gcd(t[i*2], t[i*2+1]);
		}
	}
	void build() {
		build(0, n-1, 1);
	}

	int query(int x, int y, int l, int r, int i) {
		if (l >= x && r <= y) {
			return t[i];
		}
		else if (x <= r && y >= l) {
			int m = (l + r) / 2;
			return gcd(
				query(x, y, l  , m, i*2  ),
				query(x, y, m+1, r, i*2+1));
		}
		return 0;
	}
	int query(int x, int y) {
		return query(x-1, y-1, 0, n-1, 1);
	}

	void print(int i) {
		if (i > n * 4)
			return;
		print(i * 2);
		if (s[i])
			cout << t[i] << " ";
		print(i * 2 + 1);
	}
	void print() {
		print(1);
		cout << "\n";
	}
}


const int maxA = 1e9;

struct rule {
	int x, y, z;
	// gcd(A[x:y]) == z
};

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);

	int n, m;
	cin >> n >> m;

	st1::init(n);
	st2::init(n);

	vector<rule> rules(m);
	for (auto& [x, y, z] : rules) {
		cin >> x >> y >> z;
		st1::update(x, y, z);
	}

	st2::build();

	for (auto [x, y, z] : rules) {
		int g = st2::query(x, y);
		if (g != z || g > maxA) {
			cout << "Impossible\n";
			return 0;
		}
	}
	
	st2::print();
}

File: ccc21s5.java
import java.util.*;
import java.io.*;

public class ccc21s5 {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		final int maxA = (int)1e9;
		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());

		int[] segTree1 = new int[4 * N + 1]; // gcds
		int[] segTree2 = new int[4 * N + 1]; // sequence
		boolean[] isNode = new boolean[4 * N + 1];
		Arrays.fill(segTree1, 1);

		Rule[] rules = new Rule[M];
		for (int i = 0; i < M; i++) {
			st = new StringTokenizer(br.readLine());
			int X = Integer.parseInt(st.nextToken());
			int Y = Integer.parseInt(st.nextToken());
			int Z = Integer.parseInt(st.nextToken());
			rules[i] = new Rule(X, Y, Z);
			update(segTree1, X, Y, Z, N);
		}

		build(segTree1, segTree2, isNode, N);

		for (Rule rule : rules) {
			int gcd = query(segTree2, rule.X, rule.Y, N);
			if (gcd != rule.Z || gcd > maxA) {
				System.out.println("Impossible");
				return;
			}
		}

		print(segTree2, isNode, N);
	}


	// segment tree 1
	public static void update(int[] t, int x, int y, int z, int n) {
		update(t, x-1, y-1, z, 0, n-1, 1);
	}
	public static void update(int[] t, int x, int y, int z, int l, int r, int i) {
		if (l >= x && r <= y) {
			t[i] = LCM(t[i], z);
		}
		else if (x <= r && y >= l) {
			int m = (l + r) / 2;
			update(t, x, y, z, l  , m, i*2  );
			update(t, x, y, z, m+1, r, i*2+1);
		}
	}


	// segment tree 2
	public static void build(final int[] t1, int[] t2, boolean[] s, int n) {
		build(t1, t2, s, 0, n-1, 1, 1);
	}
	public static void build(final int[] t1, int[] t2, boolean[] s, int l, int r, int i, int v) {
		v = LCM(v, t1[i]);
		if (l == r) {
			t2[i] = v;
			s[i] = true;
		}
		else {
			int m = (l + r) / 2;
			build(t1, t2, s, l  , m, i*2  , v);
			build(t1, t2, s, m+1, r, i*2+1, v);
			t2[i] = GCD(t2[i*2], t2[i*2+1]);
		}
	}


	public static int query(int[] t, int x, int y, int n) {
		return query(t, x-1, y-1, 0, n-1, 1);
	}
	public static int query(int[] t, int x, int y, int l, int r, int i) {
		if (l >= x && r <= y) {
			return t[i];
		}
		else if (x <= r && y >= l) {
			int m = (l + r) / 2;
			return GCD(
				query(t, x, y, l  , m, i*2  ),
				query(t, x, y, m+1, r, i*2+1));
		}
		return 0;
	}


	public static void print(int[] t, boolean[] s, int n) {
		print(t, s, n, 1);
		System.out.println();
	}
	public static void print(int[] t, boolean[] s, int n, int i) {
		if (i > n * 4)
			return;
		print(t, s, n, i * 2);
		if (s[i])
			System.out.print(t[i] + " ");
		print(t, s, n, i * 2 + 1);
	}


	public static int LCM(int a, int b) {
		return (a * b) / GCD(a, b);
	}

	public static int GCD(int a, int b) {
		if (b == 0) {
			return a;
		}
		else {
			return GCD(b, a % b);
		}
	}
}

class Rule {
	int X;
	int Y;
	int Z;

	public Rule(int X, int Y, int Z) {
		this.X = X;
		this.Y = Y;
		this.Z = Z;
	}
}